docs/examples.md

```diff
@@ -41,6 +41,25 @@ pipx run --spec https://github.com/psf/black/archive/18.9b0.zip black --help
 pipx run https://gist.githubusercontent.com/cs01/fa721a17a326e551ede048c5088f9e0f/raw/6bdfbb6e9c1132b1c38fdd2f195d4a24c540c324/pipx-demo.py
 ```
 
+You can also run scripts that have dependencies:
+
+If you have a script `test.py` that needs a 3rd party library like requests:
+
+```
+# test.py
+
+# Requirements:
+# requests
+
+import requests
+pipx_data = requests.get("https://pypi.org/pypi/pipx/json").json()
+print(pipx_data["info"]["version"])
+
+```
+> pipx run file:test.py
```

Comment:
I would love to add that if it's acceptable. I would suggest checking if a ile of the given name exists, if it does use that, otherwise assume a URL or package name.

It's unlikely there would be a clash between package names and filenames, as valid Python files would have to end with `.py`[^1] (we could require that).

[^1]: Although do we want to support `.pyz` and/or `.zip` as well? It may be not worth explicitly checking the extension.

Reply:
Another small improvement I'd like to add would be to support arguments for scripts. It's just a matter of adding `, *app_args` to the exec call. Would that also be acceptable, or would you prefer it as a separate PR?

Reply:
> It's unlikely there would be a clash between package names and filenames

One possible issue is that if you mistype a filename, it will be treated as a package name. For example, I was testing this and I have a script called `t.py`. I misremembered the name, and typed `pipx run test.py`. This tried to download and run [the test-py package](https://pypi.org/project/test-py/), which seems harmless, but could easily not be. There are 2352 projects on PyPI named `<something>-py`, which could clash with local files.

From [the metadata spec](https://packaging.python.org/en/latest/specifications/core-metadata/#name):

> A valid name consists only of ASCII letters and numbers, period, underscore and hyphen.

So I would suggest that:

1. Anything that doesn't match that pattern be treated as a file.
2. Anything that *does* match that pattern, but which ends in `.py`, is treated as a file.
3. Anything else (matches the pattern, doesn't end in `.py`) is treated as a file if it exists, otherwise as a package.

Filenames with directories will be caught by (1), `.py` files (whether they exist or not) by (2), and the only packages names which can't be used are ones ending in `.py` - but because of how package names are normalised, these can be referred to using `-py` as the suffix rather than `.py`.

That's a bit over-complicated, but it seems to me that it will best match what people are likely to intend.

Reply:
Grr. `urllib.parse` accepts a Windows path like `c:\temp\foo.py` as a URL, with scheme `c`. This triggers the "it's a URL" code path. So we need to check for a local file *before* checking for a URL. But again, we hit an issue with mistyped filenames. If the user types `c:\temp\fooo.py` when they mean `c:\temp\foo.py`, we want to say "invalid filename", not "unrecognised URL scheme c" 🙁

We could explicitly list the allowed URL schemes, but that feels like a regression.

I'll need to think about this some more. The good news is that, error reporting and "do what I mean" behaviour aside, processing local files as scripts is pretty straightforward.

Reply:
> One possible issue is that if you mistype a filename, it will be treated as a package name.

Good point!

> the only packages names which can't be used are ones ending in `.py`

I think some users will just follow the name displayed on PyPI and use it to install the package, so it's less likely that they will change `.py` to `-py` as the suffix. How can we handle this situation?

Reply:
Or maybe we can simply require users to use a flag(maybe `--local`/`-l`) with the local file?

Reply:
> Or maybe we can simply require users to use a flag(maybe --local/-l) with the local file?

Well, `pipx run --spec foo.py foo.py` allows the user to explicitly request a package. And every case *other* than `*.py` where the file doesn't exist, I think the behaviour is pretty clearly defined. So yeah, we could have `--local` to disambiguate, but I'd be against requiring it (just because most of the time it's unnecessary).

So if we have `--spec foo.py foo.py` for packages, `--local foo.py` for local files, and everything that doesn't conform to the syntax of a package name being treated as a URL/local file, then we only have a small area of ambiguity. If it weren't for the fact that a typo might do a costly "create a virtual environment, install a bunch of dependencies, and then run *the wrong thing*", I'd say just keep it simple and go for "if the file exists use it otherwise treat it as a package name".

Am I being over-cautious? Should we just go with that simple rule?

To be honest, I'd never have noticed the ambiguity if I hadn't manually typed in the wrong filename - if I'd used autocomplete like I normally do, it would never have completed an incorrect name, and a file in the CWD gets completed with a leading `.\`, so it wouldn't be a valid package name anyway.

I'm starting to feel that we add `--local` for disambiguation, but we treat the names of existing files as local files, and we just stop trying to be "helpful" over typos. What do you think?

> I think some users will just follow the name displayed on PyPI and use it to install the package

PyPI normalises, so the name there should always be `-py`. But of course, we don't control what's in the package docs, so people can still see the name referred to with a dot.

Reply:
> I'm starting to feel that we add --local for disambiguation, but we treat the names of existing files as local files, and we just stop trying to be "helpful" over typos. What do you think?

Yep, if users pass `--local`, then pipx will try to find the script from local and exit if it's not found. If users don't pass `--local`, it will still try to find the script from the local first. If it's not found, it will try to install a package with the normalized name.

Reply:
One further thought. Which do you prefer, `--local` or `--path`? To my mind, `pipx run --path xxx.py` reads better. Either is fine, though, we're at the bikeshedding stage now.

Reply:
I think `--path` is better actually.